src/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs

```diff
@@ -53,101 +106,173 @@ private static Delegate CreateObjectArrayDelegateRefEmit(Type delegateType, Func
-                    ilgen.Emit(OpCodes.Newarr, typeof(object));
-                }
-                ilgen.Emit(OpCodes.Stloc, argArray);
+                    bool invalidParameter = false;
+                    foreach (ParameterInfo parameter in parameters)
+                    {
+                        Type parameterType = parameter.ParameterType;
+                        if  (parameterType.IsByRefLike || parameterType.IsByRef || parameterType.IsPointer)
+                        {
+                            invalidParameter = true;
+                            break; // Don't use C# thunks for types that cannot be generic arguments
+                        }
+                    }
 
-                // populate object array
-                bool hasRefArgs = false;
-                for (int i = 0; i < parameters.Length; i++)
-                {
-                    bool paramIsByReference = parameters[i].ParameterType.IsByRef;
-                    Type paramType = parameters[i].ParameterType;
-                    if (paramIsByReference)
-                        paramType = paramType.GetElementType();
+                    if (invalidParameter)
+                        break;
 
-                    hasRefArgs = hasRefArgs || paramIsByReference;
+                    int thunkTypeArgCount = parameters.Length;
+                    if (hasReturnValue)
+                        thunkTypeArgCount++;
 
-                    ilgen.Emit(OpCodes.Ldloc, argArray);
-                    ilgen.Emit(OpCodes.Ldc_I4, i);
-                    ilgen.Emit(OpCodes.Ldarg, i + 1);
+                    Type[] thunkTypeArgs = new Type[thunkTypeArgCount];
+                    for (int i = 0; i < parameters.Length; i++)
+                    {
+                        thunkTypeArgs[i] = parameters[i].ParameterType;
+                    }
+                    MethodInfo uninstantiatedMethod;
 
-                    if (paramIsByReference)
+                    if (hasReturnValue)
                     {
-                        ilgen.Emit(OpCodes.Ldobj, paramType);
+                        thunkTypeArgs[thunkTypeArgs.Length - 1] = returnType;
+                        uninstantiatedMethod = s_FuncThunks[parameters.Length];
+                    }
+                    else
+                    {
+                        uninstantiatedMethod = s_ActionThunks[parameters.Length];
                     }
-                    Type boxType = ConvertToBoxableType(paramType);
-                    ilgen.Emit(OpCodes.Box, boxType);
-                    ilgen.Emit(OpCodes.Stelem_Ref);
-                }
 
-                if (hasRefArgs)
+                    if (thunkTypeArgs.Length > 0)
+                        s_thunks[delegateType] = thunkMethod = uninstantiatedMethod.MakeGenericMethod(thunkTypeArgs);
+                    else
+                        s_thunks[delegateType] = thunkMethod = uninstantiatedMethod;
+
+                } while (false);
```

Comment:
This `do { ... } while (false);` with `breaks` is to avoid using `goto`?

Reply:
Uh, it really should just be a helper function. And the breaks are return statements. Looks much better and is far more comprehensible.